live-examples/js-examples/proxyhandler-construct.html

```diff
@@ -0,0 +1,14 @@
+<pre>
+<code id="static-js">var proxy1 = new Proxy(function() {}, {
+  construct: function(target, argumentsList, newTarget) {
+    console.log('called: ' + argumentsList.join(', '));
+    // expected output: "called: 1"
+
+    return { value: argumentsList[0] * 10 };
+  }
+});
+
+console.log(new proxy1(1).value);
+// expected output: 10
+</code>
```

Comment:
I think it does and I love the example. A lot of the examples I wrote are quite dry with variable names such as `proxy1`, `proxy2` etc. But actually I'm in favour of adding a bit of character/humour :)

Reply:
Attempting to follow-on in the same vein - I wonder if this would make sense for `define.Property` I feel like I'm maybe missing something 🤔 

```
function setProp1() {
  var desc = { configurable: true, enumerable: true, value: 10 };
  Object.defineProperty(proxy1, 'prop1', desc);  
}

var handler1 = {
  defineProperty: function(target, prop, descriptor) {
    console.log('called: ' + prop);
    // expected output: "called: prop1"

    return true;
  }  
}

var proxy1 = new Proxy(setProp1, handler1);
new proxy1();
```

Reply:
Similarly the `deleteProperty` example would look like this? :

```
var handler1 = {
  deleteProperty: function(target, prop) {
    if (prop in target){
      delete target[prop];

      console.log('property removed: ' + prop);
      // expected output: "property removed: prop1"
    }
  }  
}

var proxy1 = new Proxy({}, handler1);

proxy1.prop1 = 10;

console.log('prop1' in proxy1);
// expected output: true

result = delete proxy1.prop1;

console.log('prop1' in proxy1);
// expected output: false
```

Reply:
I think it's helpful with these examples to think of use cases - "why would someone want to do a thing like this". The Google article (https://developers.google.com/web/updates/2016/02/es2015-proxies#use_cases) has some good use cases. From what I understand there are a few obvious use cases for proxies. For example:

* logging particular operations on an object: that's what the Monster constructor one does. Here the handler just logs that the call happened, then forwards it on to the target.

* just blocking a particular operation entirely, for example blocking a `delete`.

* conditionally blocking an operation. I thought the example here was a nice one: https://ponyfoo.com/articles/es6-proxies-in-depth#set - it traps `get` and `set` to prevent access to properties whose name starts with an underscore. Or validating an operation in some other way. For example, maybe the monster has to have an even number of eyes:

```
function Monster() {
  this.eyeCount = 4;
}

var handler = {
  set: function(obj, prop, value) {
    if ((prop === 'eyeCount') && ((value % 2) !== 0)) {
      console.log("Monsters must have an even number of eyes");
    } else {
      return Reflect.set(...arguments);
    }
  }
};

var monster = new Monster();
var monsterProxy = new Proxy(monster, handler);
monsterProxy.eyeCount = 5;
// expected output: "Monsters must have an even number of eyes"
console.log(monsterProxy.eyeCount);
// expected output: 4
monsterProxy.eyeCount = 8;
console.log(monsterProxy.eyeCount);
// expected output: 5
```
I struggle a bit with examples where the target is an empty object, because I don't see the point  in proxying to nothing.
 
For the `defineProperty` example, there's a nice example here: https://ponyfoo.com/articles/es6-proxy-traps-in-depth#it-s-a-trap that uses it to prevent you from adding new private (underscore-prefixed) properties.

One problem is that these examples are definitely more than 12 lines long, but I think maybe in this case we just need 20 or so lines to get a good example.

For the `deleteProperty` example, it's nice IMO to have a named target, so you can show that the delete has an effect on the target, not just on the proxy. For example:

```
function Monster() {
  this.texture = "scaly";
}

var handler = {
  deleteProperty: function(target, prop) {
    if (prop in target) {
      delete target[prop];
      console.log('property removed: ' + prop);
      // expected output: "property removed: texture"
    }
  }  
}

var monster = new Monster();
console.log(monster.texture); 

var monsterProxy = new Proxy(monster, handler);
delete monsterProxy.texture;
console.log(monster.texture);    // note, monster not monsterProxy: the delete affects the target
```